common/src/main/java/com/microsoft/identity/common/internal/providers/oauth2/OAuth2Strategy.java

```diff
@@ -230,6 +245,101 @@ protected final void setAuthorizationEndpoint(final String authorizationEndpoint
         mAuthorizationEndpoint = authorizationEndpoint;
     }
 
+    public AuthorizationResult getDeviceCode(MicrosoftStsAuthorizationRequest authorizationRequest, @Nullable String authorityUrl) throws IOException {
+        final String methodName = ":getDeviceCode";
+
+        // Set up connection
+        String urlBody;
+
+        if (authorityUrl == null) {
+            // Fetch the tenant from the authority
+            String tenant = authorizationRequest.getAuthority().getPath().split("/")[1];
```

Comment:
We need to pull this logic into the `MicrosoftStsOAuth2Configuration` -- I don't think we want to extract the tenant info like this from the URL, but instead we should inspect the `PublicClientApplicationConfiguration` to get the audience/tenant.

Hazard from parsing it from the URL is that this basically assumes the URL structure will never change. For AAD today, that's a safe assumption -- if we add support for direct-ADFS auth later though, this will break

Reply:
Parsing is definitely not ideal. I wanted to get a working PR out, I couldn't find the tenant as it's own field in the request's authority. Hopefully this will be fixed with the inclusion of `getDeviceAuthorizationEndpoint()` in `MicrosoftStsOAuth2Configuration`.

Reply:
On a side note, we should be able to just append /oauth2/v2.0/devicecode to the authority URL (getAuthority() inside MicrosoftStsOAuth2Configuration - see getTokenEndpoint() and getEndpointUrlFromRootAndSuffix() for more info)

Reply:
the Authority is the same in the `authorizationRequest` and the `mConfig`. The issue is that `path` (which contains tenant) ends with `/common`, so it's structured like `/{tenant}/common`. For DCF, I need to remove the `/common` at the end of it before adding `/oauth2/v2.0/devicecode` or else the request will return 404 "not found".

Reply:
Same thing occurs in the token polling request, but I can just change `mTokenEndpoint` there and the whole process is a lot less messy.

Reply:
Would you mind sharing your config file? I'd expect either /{tenantID} or /common, /organization, etc.

Reply:
Git doesn't like Json files? Here's a screenshot:
![Annotation 2020-07-29 123118](https://user-images.githubusercontent.com/65684798/88844383-6c921000-d197-11ea-96d1-0d0e92633688.jpg)

Reply:
At some point, /common is being added to the tenant in `authority_url`

Reply:
That sounds like a bug to me....

Reply:
iirc that value would be the tenant and /common at the end of it

Reply:
i'm assuming you're referring to the `mConfig` object in `OAuth2Strategy`?

Reply:
Yes, but I'm referring to the class itself. @iambmelt's original suggestion is to put this url logic in that class and get the endpoint from there.

Reply:
@t-fadura AFAIK, DCF is only supported for AAD and not B2C. So you should be able to get the `tenantId` as follows:

`((AzureActiveDirectoryAuthority) mConfig.getDefaultAuthority()).getAudience().getTenantId();`

Reply:
I have an `AzureActiveDirectoryAuthority` object in `LocalMsalController` being passed through command parameters. `mTenantId` in it is "common", where as `mCloudUrl` in it is "https://login.microsoftonline.com/{MyTenant}". Maybe this is the base issue?

Reply:
@rpdome @shahzaib I found one issue, my methods were missing this line
`Authority.KnownAuthorityResult authorityResult = Authority.getKnownAuthorityResult(parametersWithScopes.getAuthority());`

This line would implicitly call `performCloudDiscovery()` in `AzureActiveDirectory`, which plays a big part in how the authorities are structured since `mAzureActiveDirectoryCloud` is no longer null in `AzureActiveDirectoryAuthority`. (Sorry for the confusing explanation)

Currently tying to figure out why "common" is in the mTenantId field. Once this is fixed there shouldn't been any need to mess with the token endpoint

Reply:
Authority issue fixed, I was not passing the correct authority object to the command parameters object. Resolving after fix is commited.